Events History with Tests

zerox1212 opened the pull request:
This commit is very functional, but WhereClause is not supported yet.
Default history will fail tests because it raises NotImplemented on
event calls.

zerox1212 commented:
Can we please look at reviewing and merging this? I did everything possible to make it complete, including tests.

There are getting to be so many changes that if you change the master and I have to rebase it's going to be very difficult to fix and I might have to stop working...

oroulet commented on opcua/common/subscription.py:
```diff
@@ -52,10 +52,21 @@ def __str__(self):
+            if not key.startswith("__") and key is not "server_handle":
+                if key is "SourceName":
+                    # hack because client is expecting string not byte string?
+                    field_vars[key] = ua.Variant(value.decode(encoding='utf-8'))
```
remove the hack. will get fixed somewhere else

zerox1212 replied:
This is done.

oroulet commented on opcua/common/subscription.py:
```diff
@@ -52,10 +52,21 @@ def __str__(self):
         return "EventResult({})".format([str(k) + ":" + str(v) for k, v in self.__dict__.items()])
     __repr__ = __str__
 
+    def get_field_variants(self):
```
can you give that method a more understanble name? since it is now exposed. get_event_properties_as_field_list() or whatever make sense

zerox1212 replied:
This is done. It is now named `get_event_props_as_field_dict`.

oroulet commented on opcua/server/history.py:
```diff
@@ -162,7 +163,7 @@ def _create_subscription(self, handler):
         return Subscription(self.iserver.isession, params, handler)
 
-    def historize(self, node, period=timedelta(days=7), count=0):
+    def historize_var(self, node, period=timedelta(days=7), count=0):
```
histprize_data_change would be a better name since this is what we use at other places

zerox1212 replied:
This is done.

oroulet commented on opcua/server/internal_server.py:
```diff
@@ -153,16 +153,16 @@ def register_server2(self, params):
         return InternalSession(self, self.aspace, self.subscription_service, name, user=user, external=external)
 
-    def enable_history(self, node, period=timedelta(days=7), count=0):
+    def enable_history_var(self, node, period=timedelta(days=7), count=0):
```
same here: enable_history_data_change

zerox1212 replied:
This is done.

oroulet commented on opcua/server/internal_subscription.py:
```diff
@@ -91,7 +91,11 @@ def _create_monitored_item(self, params):
+                if ev_notify_byte is not None:
+                    if ev_notify_byte & 1 == 0:
+                        result.StatusCode = ua.StatusCode(ua.StatusCodes.BadServiceUnsupported)
+                else:
```
if ev_notify_byte is not None or ev_notify_byte & 1 == 0:
           result.StatusCode = ua.StatusCode(ua.StatusCodes.BadServiceUnsupported)

that is the same and muvh simpler

zerox1212 replied:
This is done.

oroulet commented:
I am fine with merging this. but we need to fix the tests before merging. I think the simplest thing to do now is to move your new tests to TestHistorySQL class then they will only be executed with sql backend

zerox1212 commented:
I will make the changes you listed above tomorrow. How can I move the event tests to only the SQL class?

zerox1212 commented:
>I am fine with merging this. but we need to fix the tests before merging. I think the simplest thing to do now is to move your new tests to TestHistorySQL class then they will only be executed with sql backend

Would you consider making the ":memory:" SQL history the default and removing the python dict? Seems like it makes more sense to just maintain one implementation. Both are in memory so I don't think there is a performance impact.

zerox1212 commented:
Cleaned up the implementation as you noted. I also moved event tests to TestHistorySQL class. If you can merge it, I can start adding support for WhereClause.

zerox1212 commented:
I think everything is fixed now. When event subscription supports the WhereClause I can add it to history as well.

The only minor complaint from myself is that the SQL code for historizing data values could be cleaned up to better match the events implementation for inserting/reading data. For now I think it's good enough.

Maybe next we can add alarms. :)

oroulet commented:
I did not have time to check everything but I merge since it does not seem to break anything. We can fix small things late.

oroulet commented:
>> Would you consider making the ":memory:" SQL history the default and removing the python dict?

Not sure what to do. My plan was to have the dict stuff as default as it does not require any dependencies but make the examples use the sql implementation AND a file as users of history will want a file.
Maintaining 2 implementation has the advantage to oblige us to thing about code re-use,  users will want to make their own storage class. To share as much code as possible between our implementationsm maybe we should modify th existing interface...

>> One other note, I don't think event history will react very well to adding custom event fields to a node that is already historized. All the custom fields need to be added before enable history is called.

This is fine. but document it!
